File: go/signer.go
package clientattestation

import (
	"bytes"
	"crypto"
	"crypto/ecdsa"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/lestrrat-go/jwx/v2/jwk"
)

// JwsSigner is a JWS signing capability whose private key may live outside the process (a vault, HSM, or
// KMS). Sign returns the RAW JWS signature over the signing input — for ECDSA the fixed-width r||s
// concatenation RFC 7515 requires, not ASN.1/DER. *SigningKeyPair covers the local-key case;
// *OpenBaoTransitSigner signs inside a vault.
type JwsSigner interface {
	Algorithm() string
	KeyID() string
	PublicJWK() map[string]any
	Sign(signingInput []byte) ([]byte, error)
}

// signExternal assembles a compact JWS whose signature is produced by an external signer. The header
// carries the signer's kid (external signers hold issuing keys, referenced by id, never an embedded jwk).
func signExternal(claims map[string]any, signer JwsSigner, typ string) (string, error) {
	headerJSON, err := json.Marshal(map[string]any{"alg": signer.Algorithm(), "typ": typ, "kid": signer.KeyID()})
	if err != nil {
		return "", err
	}
	payloadJSON, err := json.Marshal(claims)
	if err != nil {
		return "", fmt.Errorf("unable to marshal claims: %w", err)
	}
	enc := base64.RawURLEncoding
	signingInput := enc.EncodeToString(headerJSON) + "." + enc.EncodeToString(payloadJSON)
	sig, err := signer.Sign([]byte(signingInput))
	if err != nil {
		return "", err
	}
	return signingInput + "." + enc.EncodeToString(sig), nil
}

var keyTypeAlg = map[string]string{"ecdsa-p256": "ES256", "ecdsa-p384": "ES384", "ecdsa-p521": "ES512"}
var algHash = map[string]string{"ES256": "sha2-256", "ES384": "sha2-384", "ES512": "sha2-512"}

// OpenBaoTransitSigner is a JwsSigner backed by an OpenBao / HashiCorp Vault transit engine: the attestation
// is signed inside the vault (POST /v1/transit/sign/<key> with marshaling_algorithm=jws, which returns the
// fixed-width r||s) and the private key never leaves it. NewOpenBaoTransitSigner reads the key metadata to
// pin the latest version, derive the public JWK and compute its RFC 7638 kid. Fail-closed.
type OpenBaoTransitSigner struct {
	base       string
	token      string
	keyName    string
	algorithm  string
	hash       string
	keyVersion int64
	keyID      string
	publicJWK  map[string]any
	client     *http.Client
}

// NewOpenBaoTransitSigner builds a signer for the transit key keyName, reachable at baoAddr with token.
func NewOpenBaoTransitSigner(baoAddr, token, keyName string) (*OpenBaoTransitSigner, error) {
	s := &OpenBaoTransitSigner{
		base:    strings.TrimRight(baoAddr, "/"),
		token:   token,
		keyName: keyName,
		client:  &http.Client{Timeout: 5 * time.Second},
	}
	data, err := s.request(http.MethodGet, "/v1/transit/keys/"+keyName, nil)
	if err != nil {
		return nil, err
	}
	keyType, _ := data["type"].(string)
	alg, ok := keyTypeAlg[keyType]
	if !ok {
		return nil, fmt.Errorf("transit key %q has unsupported type for JWS signing: %s", keyName, keyType)
	}
	s.algorithm = alg
	s.hash = algHash[alg]
	s.keyVersion = int64(toFloat(data["latest_version"]))
	keys, _ := data["keys"].(map[string]any)
	latest, _ := keys[strconv.FormatInt(s.keyVersion, 10)].(map[string]any)
	pemStr, _ := latest["public_key"].(string)
	jwkMap, kid, err := ecJWKFromPEM(pemStr)
	if err != nil {
		return nil, err
	}
	jwkMap["kid"] = kid
	jwkMap["alg"] = alg
	s.publicJWK = jwkMap
	s.keyID = kid
	return s, nil
}

// Algorithm returns the JWS "alg" this signer produces.
func (s *OpenBaoTransitSigner) Algorithm() string { return s.algorithm }

// KeyID returns the RFC 7638 thumbprint kid of the transit key.
func (s *OpenBaoTransitSigner) KeyID() string { return s.keyID }

// PublicJWK returns a fresh copy of the public-only JWK (including "kid").
func (s *OpenBaoTransitSigner) PublicJWK() map[string]any {
	out := make(map[string]any, len(s.publicJWK))
	for k, v := range s.publicJWK {
		out[k] = v
	}
	return out
}

// Sign signs the JWS signing input inside the vault, returning the raw r||s signature.
func (s *OpenBaoTransitSigner) Sign(signingInput []byte) ([]byte, error) {
	body, err := json.Marshal(map[string]any{
		"input":                base64.StdEncoding.EncodeToString(signingInput),
		"marshaling_algorithm": "jws",
		"hash_algorithm":       s.hash,
		"key_version":          s.keyVersion,
	})
	if err != nil {
		return nil, err
	}
	data, err := s.request(http.MethodPost, "/v1/transit/sign/"+s.keyName, body)
	if err != nil {
		return nil, err
	}
	signature, _ := data["signature"].(string)
	if signature == "" {
		return nil, fmt.Errorf("transit sign response carried no signature")
	}
	raw := signature[strings.LastIndex(signature, ":")+1:] // envelope: vault:v<n>:<base64url(r||s)>
	return base64.RawURLEncoding.DecodeString(raw)
}

func (s *OpenBaoTransitSigner) request(method, path string, body []byte) (map[string]any, error) {
	var reader io.Reader
	if body != nil {
		reader = bytes.NewReader(body)
	}
	req, err := http.NewRequest(method, s.base+path, reader)
	if err != nil {
		return nil, err
	}
	req.Header.Set("X-Vault-Token", s.token)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	resp, err := s.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("OpenBao unreachable at %s: %w", s.base, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("OpenBao returned HTTP %d for %s", resp.StatusCode, path)
	}
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var parsed struct {
		Data map[string]any `json:"data"`
	}
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return nil, fmt.Errorf("unparseable OpenBao response: %w", err)
	}
	if parsed.Data == nil {
		return nil, fmt.Errorf("OpenBao response carried no data")
	}
	return parsed.Data, nil
}

// ecJWKFromPEM converts a transit public_key PEM (SubjectPublicKeyInfo) into an EC public JWK plus its
// RFC 7638 thumbprint kid.
func ecJWKFromPEM(pemStr string) (map[string]any, string, error) {
	block, _ := pem.Decode([]byte(pemStr))
	if block == nil {
		return nil, "", fmt.Errorf("transit key metadata carried no valid PEM public key")
	}
	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, "", fmt.Errorf("unparseable transit public key: %w", err)
	}
	if _, ok := pub.(*ecdsa.PublicKey); !ok {
		return nil, "", fmt.Errorf("transit public key is not ECDSA")
	}
	key, err := jwk.FromRaw(pub)
	if err != nil {
		return nil, "", fmt.Errorf("unable to import transit public key: %w", err)
	}
	jwkMap, err := canonicalPublicJWK(key)
	if err != nil {
		return nil, "", err
	}
	tp, err := key.Thumbprint(crypto.SHA256)
	if err != nil {
		return nil, "", fmt.Errorf("unable to compute transit key thumbprint: %w", err)
	}
	return jwkMap, base64.RawURLEncoding.EncodeToString(tp), nil
}

func toFloat(v any) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case json.Number:
		f, _ := n.Float64()
		return f
	}
	return 0
}

File: typescript/src/signer.ts
/**
 * External JWS signing — keys that live outside the process (a vault, HSM, or KMS).
 *
 * {@link SigningKeyPair} covers the local-key case; {@link OpenBaoTransitSigner} signs inside an
 * OpenBao / HashiCorp Vault transit engine so the attester's private key never leaves it.
 */
import { calculateJwkThumbprint, exportJWK, importSPKI, type JWK } from "jose";

const b64url = (bytes: Uint8Array): string => Buffer.from(bytes).toString("base64url");

/** A JWS signing capability whose private key may live outside the process. `sign` returns the RAW JWS
 * signature over the signing input — for ECDSA the fixed-width `r||s`, not ASN.1/DER. */
export interface JwsSigner {
  readonly algorithm: string;
  readonly keyId: string;
  publicJwk(): JWK;
  sign(signingInput: Uint8Array): Promise<Uint8Array>;
}

/** Assemble a compact JWS whose signature is produced by an external signer (header carries `kid`; external
 * signers hold issuing keys, referenced by id, never an embedded jwk). */
export async function signExternal(
  claims: Record<string, unknown>,
  signer: JwsSigner,
  typ: string,
): Promise<string> {
  const enc = new TextEncoder();
  const header = { alg: signer.algorithm, typ, kid: signer.keyId };
  const signingInput = `${b64url(enc.encode(JSON.stringify(header)))}.${b64url(enc.encode(JSON.stringify(claims)))}`;
  const signature = await signer.sign(enc.encode(signingInput));
  return `${signingInput}.${b64url(signature)}`;
}

const KEY_TYPE_ALG: Record<string, string> = { "ecdsa-p256": "ES256", "ecdsa-p384": "ES384", "ecdsa-p521": "ES512" };
const ALG_HASH: Record<string, string> = { ES256: "sha2-256", ES384: "sha2-384", ES512: "sha2-512" };

/**
 * A {@link JwsSigner} backed by an OpenBao / HashiCorp Vault transit engine: the attestation is signed
 * inside the vault (`POST /v1/transit/sign/<key>` with `marshaling_algorithm=jws`, which returns the
 * fixed-width `r||s`) and the private key never leaves it. `create` reads the key metadata to pin the
 * latest version, derive the public JWK and compute its RFC 7638 `kid`. Fail-closed (vault errors throw).
 */
export class OpenBaoTransitSigner implements JwsSigner {
  readonly algorithm: string;
  readonly keyId: string;
  readonly #base: string;
  readonly #token: string;
  readonly #keyName: string;
  readonly #hash: string;
  readonly #keyVersion: number;
  readonly #publicJwk: JWK;

  private constructor(
    base: string, token: string, keyName: string, algorithm: string,
    hash: string, keyVersion: number, publicJwk: JWK, keyId: string,
  ) {
    this.#base = base;
    this.#token = token;
    this.#keyName = keyName;
    this.algorithm = algorithm;
    this.#hash = hash;
    this.#keyVersion = keyVersion;
    this.#publicJwk = publicJwk;
    this.keyId = keyId;
  }

  static async create(baoAddr: string, token: string, keyName: string): Promise<OpenBaoTransitSigner> {
    const base = baoAddr.replace(/\/+$/, "");
    const data = await request("GET", base, token, `/v1/transit/keys/${keyName}`);
    const keyType = String(data.type);
    const algorithm = KEY_TYPE_ALG[keyType];
    if (!algorithm) {
      throw new Error(`transit key '${keyName}' has unsupported type for JWS signing: ${keyType}`);
    }
    const keyVersion = Number(data.latest_version);
    const keys = data.keys as Record<string, { public_key: string }>;
    const pem = keys[String(keyVersion)].public_key;
    const jwk = await exportJWK(await importSPKI(pem, algorithm, { extractable: true }));
    const keyId = await calculateJwkThumbprint(jwk, "sha256");
    const publicJwk: JWK = { ...jwk, kid: keyId, alg: algorithm };
    return new OpenBaoTransitSigner(base, token, keyName, algorithm, ALG_HASH[algorithm], keyVersion, publicJwk, keyId);
  }

  publicJwk(): JWK {
    return { ...this.#publicJwk };
  }

  async sign(signingInput: Uint8Array): Promise<Uint8Array> {
    const body = JSON.stringify({
      input: Buffer.from(signingInput).toString("base64"),
      marshaling_algorithm: "jws",
      hash_algorithm: this.#hash,
      key_version: this.#keyVersion,
    });
    const data = await request("POST", this.#base, this.#token, `/v1/transit/sign/${this.#keyName}`, body);
    const signature = data.signature as string | undefined;
    if (!signature) {
      throw new Error("transit sign response carried no signature");
    }
    const raw = signature.slice(signature.lastIndexOf(":") + 1); // envelope: vault:v<n>:<base64url(r||s)>
    return new Uint8Array(Buffer.from(raw, "base64url"));
  }
}

async function request(
  method: string, base: string, token: string, path: string, body?: string,
): Promise<Record<string, unknown>> {
  let response: Response;
  try {
    response = await fetch(base + path, {
      method,
      headers: { "X-Vault-Token": token, ...(body ? { "Content-Type": "application/json" } : {}) },
      body,
    });
  } catch (exc) {
    throw new Error(`OpenBao unreachable at ${base}: ${exc instanceof Error ? exc.message : String(exc)}`);
  }
  if (response.status !== 200) {
    throw new Error(`OpenBao returned HTTP ${response.status} for ${path}`);
  }
  const parsed = (await response.json()) as { data?: Record<string, unknown> };
  if (!parsed.data) {
    throw new Error("OpenBao response carried no data");
  }
  return parsed.data;
}
